snuba/query/expressions.py

```diff
@@ -0,0 +1,183 @@
+from __future__ import annotations
+
+from abc import ABC, abstractmethod
+from dataclasses import dataclass
+from typing import Callable, Iterable, Iterator, Optional, Sequence
+
+from snuba.query.collections import NodeContainer
+
+
+class Expression(ABC):
+    """
+    A node in the Query AST. This can be a leaf or an intermediate node.
+    It also represents an expression that can be resolved to a value. This
+    includes column names, function calls and boolean conditions (which are
+    function calls themselves in the AST).
+
+    The root of the tree is not a Node itself yet (since it is the Query object).
+    Representing the root as a node itself does not seem very useful right now
+    since we never traverse the full tree. We could revisit that later.
+    """
+
+    @abstractmethod
+    def format(self) -> str:
+        """
+        Turn this node into a string for the Clickhouse query.
+
+        TODO: provide a clickhouse formatter to this method so that, through
+        a strategy pattern, this class will provide the content for the query
+        and the clickhouse formatter will provide the format.
+        """
+        raise NotImplementedError
+
+
+class ExpressionContainer(NodeContainer[Expression]):
+    """
+    Container able to iterate over expressions and to transform them in place.
+    This class exists for two reasons:
+    - in some place we check the type of an object through isinstance, and, since
+      the parameter of a generic disappears at runtime we cannot do something like
+      isinstance(a, NodeContainer[Expression])
+    - provides some common feature to transform and iterate over the children of
+      a hierarchical expression like a tree.
+    """
+
+    def _iterate_over_children(self,
+        children: Iterable[Expression],
+    ) -> Iterator[Expression]:
+        """
+        Traverses the children of a hierarchical container like a tree.
+        """
+        for child in children:
+            if isinstance(child, ExpressionContainer):
+                for sub in child:
+                    yield sub
+            else:
+                yield child
+
+    def _transform_children(self,
+        children: Sequence[Expression],
+        func: Callable[[Expression], Expression],
+    ) -> Sequence[Expression]:
+        """
+        Transforms in place the children of a hierarchical node by applying
+        a mapping function.
+        """
+        def process_child(param: Expression) -> Expression:
+            r = func(param)
+            if r == param and isinstance(r, ExpressionContainer):
+                # The expression was not replaced by the function, which
+                # means it was unchanged. This means we need to traverse
+                # its children.
+                r.transform(func)
+            return r
+
+        return list(map(process_child, children))
+
+
+@dataclass
+class AliasedExpression(Expression, ExpressionContainer):
```

Comment:
I think ClickHouse might allow you to alias *any* arbitrary expression, so I'm wondering if adding the alias might be better on the expression after all:

```
SELECT 
    a, 
    b, 
    c
FROM system.one 
GROUP BY dummy AS a
HAVING a = 0 AS b
ORDER BY b AS c ASC

┌─a─┬─b─┬─c─┐
│ 0 │ 1 │ 1 │
└───┴───┴───┘

1 rows in set. Elapsed: 0.013 sec. 
```

I'm not entirely sure how all of these aliases are resolved though, so not sure how much sense this makes generically.

Reply:
In the current abstraction, where everything ended up being an expression, I don't think it makes much difference. The only trick is manipulating aliases, which is marginally simpler if the alias is an attribute: 
Imagine this case: I have an expression like this f(A) AS B and I want to transform it into g(x) as B.

If the alias is an attribute:
I will have a node f(A) as B and I just replace it with a node g(x) with the same alias. Simple

If the alias is a Wrapper:
I would have a node f(A) wrapped by an Alias B. Now When I transform f(A) into g(X) I do not have access to the alias since it is a parent node, which means I will end into an Alias node A wrapping an alias node with alias f(A) which wraps G(X) (if we replace node A we swap it with node B with alias A to preserve the naming). 
Which means at the end of the process we need to collapse aliases. It is feasible and it would work, it is just a little more complex.